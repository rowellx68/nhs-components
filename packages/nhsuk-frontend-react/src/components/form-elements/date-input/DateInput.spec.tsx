import React, { act } from 'react';
import { it, expect, vi } from 'vitest';
import { fireEvent, render } from '@testing-library/react';
import { composeStory } from '@storybook/react';
import meta, {
  Default as DefaultStory,
  WithError as WithErrorStory,
  WithErrorObject as WithErrorObjectStory,
} from './DateInput.stories';

const Default = composeStory(DefaultStory, meta);
const WithError = composeStory(WithErrorStory, meta);
const WithErrorObject = composeStory(WithErrorObjectStory, meta);

it('should render the date input component', () => {
  const { container } = render(<Default />);

  expect(container.querySelectorAll('input')).toHaveLength(3);
  expect(container).toMatchSnapshot();
});

it('should render the date input component with a value', () => {
  const { container, getByLabelText } = render(
    <Default value={{ day: '1', month: '1', year: '2000' }} />,
  );

  expect(container.querySelectorAll('input')).toHaveLength(3);
  expect(getByLabelText('Day')).toHaveValue('1');
  expect(getByLabelText('Month')).toHaveValue('1');
  expect(getByLabelText('Year')).toHaveValue('2000');
});

it('should render the date input component with a label and hint', () => {
  const { container, getByText } = render(
    <Default label="Date of birth" hint="For example, 31 3 1980" />,
  );

  expect(getByText('Date of birth')).toBeInTheDocument();
  expect(getByText('For example, 31 3 1980')).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

it('should render the date input component with an error message', () => {
  const { container, getByText } = render(<WithError />);

  expect(getByText('Please enter a valid date of birth')).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

it('should render the date input component with an error object', () => {
  const { container } = render(<WithErrorObject />);

  expect(container).toHaveTextContent(
    'Please enter a valid day. Please enter a valid month.',
  );
  expect(container).toMatchSnapshot();
});

it('should call the onChange function when the input is changed', () => {
  const onChange = vi.fn();
  const { getByLabelText } = render(<Default onChange={onChange} />);

  act(() => {
    fireEvent.input(getByLabelText('Day'), { target: { value: '1' } });
    fireEvent.input(getByLabelText('Month'), { target: { value: '1' } });
    // simulate typing user input
    fireEvent.input(getByLabelText('Year'), { target: { value: '2' } });
    fireEvent.input(getByLabelText('Year'), { target: { value: '20' } });
    fireEvent.input(getByLabelText('Year'), { target: { value: '200' } });
    fireEvent.input(getByLabelText('Year'), { target: { value: '2000' } });
  });

  expect(onChange).toHaveBeenCalledTimes(6);
});
