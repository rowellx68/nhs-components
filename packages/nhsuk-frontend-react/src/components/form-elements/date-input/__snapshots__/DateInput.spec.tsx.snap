// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render the date input component 1`] = `
<div>
  <div
    class="nhsuk-form-group"
  >
    <div
      class="nhsuk-form-group"
    >
      <fieldset
        aria-describedby="dateinput_:r0:_:r1:--hint"
        aria-labelledby="dateinput_:r0:--label"
        class="nhsuk-fieldset"
        role="group"
      >
        <legend
          class="nhsuk-legend"
          for="dateinput_:r0:"
          id="dateinput_:r0:--label"
        >
          Date of birth
        </legend>
        <div
          class="nhsuk-hint"
          id="dateinput_:r0:_:r1:--hint"
        >
          For example, 31 3 1980
        </div>
        <div
          class="nhsuk-date-input"
          id="dateinput_:r0:"
        >
          <div
            class="nhsuk-date-input__item"
          >
            <div
              class="nhsuk-form-group"
            >
              <label
                class="nhsuk-label nhsuk-date-input__label"
                for="dateinput_:r0:-day"
                id="dateinput_:r0:-day--label"
              >
                Day
              </label>
              <input
                aria-labelledby="dateinput_:r0:-day--label"
                as="input"
                class="nhsuk-input nhsuk-input--width-2 nhsuk-date-input__input"
                id="dateinput_:r0:-day"
                inputmode="numeric"
                name="dateinput_:r0:-day"
                pattern="[0-9]*"
                type="text"
                value="31"
              />
            </div>
          </div>
          <div
            class="nhsuk-date-input__item"
          >
            <div
              class="nhsuk-form-group"
            >
              <label
                class="nhsuk-label nhsuk-date-input__label"
                for="dateinput_:r0:-month"
                id="dateinput_:r0:-month--label"
              >
                Month
              </label>
              <input
                aria-labelledby="dateinput_:r0:-month--label"
                as="input"
                class="nhsuk-input nhsuk-input--width-2 nhsuk-date-input__input"
                id="dateinput_:r0:-month"
                inputmode="numeric"
                name="dateinput_:r0:-month"
                pattern="[0-9]*"
                type="text"
                value="3"
              />
            </div>
          </div>
          <div
            class="nhsuk-date-input__item"
          >
            <div
              class="nhsuk-form-group"
            >
              <label
                class="nhsuk-label nhsuk-date-input__label"
                for="dateinput_:r0:-year"
                id="dateinput_:r0:-year--label"
              >
                Year
              </label>
              <input
                aria-labelledby="dateinput_:r0:-year--label"
                as="input"
                class="nhsuk-input nhsuk-input--width-4 nhsuk-date-input__input"
                id="dateinput_:r0:-year"
                inputmode="numeric"
                name="dateinput_:r0:-year"
                pattern="[0-9]*"
                type="text"
                value="1980"
              />
            </div>
          </div>
        </div>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`should render the date input component with a label and hint 1`] = `
<div>
  <div
    class="nhsuk-form-group"
  >
    <div
      class="nhsuk-form-group"
    >
      <fieldset
        aria-describedby="dateinput_:rs:_:rt:--hint"
        aria-labelledby="dateinput_:rs:--label"
        class="nhsuk-fieldset"
        role="group"
      >
        <legend
          class="nhsuk-legend"
          for="dateinput_:rs:"
          id="dateinput_:rs:--label"
        >
          Date of birth
        </legend>
        <div
          class="nhsuk-hint"
          id="dateinput_:rs:_:rt:--hint"
        >
          For example, 31 3 1980
        </div>
        <div
          class="nhsuk-date-input"
          id="dateinput_:rs:"
        >
          <div
            class="nhsuk-date-input__item"
          >
            <div
              class="nhsuk-form-group"
            >
              <label
                class="nhsuk-label nhsuk-date-input__label"
                for="dateinput_:rs:-day"
                id="dateinput_:rs:-day--label"
              >
                Day
              </label>
              <input
                aria-labelledby="dateinput_:rs:-day--label"
                as="input"
                class="nhsuk-input nhsuk-input--width-2 nhsuk-date-input__input"
                id="dateinput_:rs:-day"
                inputmode="numeric"
                name="dateinput_:rs:-day"
                pattern="[0-9]*"
                type="text"
                value="31"
              />
            </div>
          </div>
          <div
            class="nhsuk-date-input__item"
          >
            <div
              class="nhsuk-form-group"
            >
              <label
                class="nhsuk-label nhsuk-date-input__label"
                for="dateinput_:rs:-month"
                id="dateinput_:rs:-month--label"
              >
                Month
              </label>
              <input
                aria-labelledby="dateinput_:rs:-month--label"
                as="input"
                class="nhsuk-input nhsuk-input--width-2 nhsuk-date-input__input"
                id="dateinput_:rs:-month"
                inputmode="numeric"
                name="dateinput_:rs:-month"
                pattern="[0-9]*"
                type="text"
                value="3"
              />
            </div>
          </div>
          <div
            class="nhsuk-date-input__item"
          >
            <div
              class="nhsuk-form-group"
            >
              <label
                class="nhsuk-label nhsuk-date-input__label"
                for="dateinput_:rs:-year"
                id="dateinput_:rs:-year--label"
              >
                Year
              </label>
              <input
                aria-labelledby="dateinput_:rs:-year--label"
                as="input"
                class="nhsuk-input nhsuk-input--width-4 nhsuk-date-input__input"
                id="dateinput_:rs:-year"
                inputmode="numeric"
                name="dateinput_:rs:-year"
                pattern="[0-9]*"
                type="text"
                value="1980"
              />
            </div>
          </div>
        </div>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`should render the date input component with an error message 1`] = `
<div>
  <div
    class="nhsuk-form-group nhsuk-form-group--error"
  >
    <div
      class="nhsuk-form-group"
    >
      <fieldset
        aria-describedby="dateinput_:r1a:_:r1b:--hint dateinput_:r1a:--error-message"
        aria-labelledby="dateinput_:r1a:--label"
        class="nhsuk-fieldset"
        role="group"
      >
        <legend
          class="nhsuk-legend"
          for="dateinput_:r1a:"
          id="dateinput_:r1a:--label"
        >
          Date of birth
        </legend>
        <div
          class="nhsuk-hint"
          id="dateinput_:r1a:_:r1b:--hint"
        >
          For example, 31 3 1980
        </div>
        <span
          class="nhsuk-error-message"
          id="dateinput_:r1a:--error-message"
        >
          <span
            class="nhsuk-u-visually-hidden"
          >
            Error: 
          </span>
          Please enter a valid date of birth
        </span>
        <div
          class="nhsuk-date-input"
          id="dateinput_:r1a:"
        >
          <div
            class="nhsuk-date-input__item"
          >
            <div
              class="nhsuk-form-group"
            >
              <label
                class="nhsuk-label nhsuk-date-input__label"
                for="dateinput_:r1a:-day"
                id="dateinput_:r1a:-day--label"
              >
                Day
              </label>
              <input
                aria-labelledby="dateinput_:r1a:-day--label"
                as="input"
                class="nhsuk-input nhsuk-input--width-2 nhsuk-date-input__input nhsuk-input--error"
                id="dateinput_:r1a:-day"
                inputmode="numeric"
                name="dateinput_:r1a:-day"
                pattern="[0-9]*"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="nhsuk-date-input__item"
          >
            <div
              class="nhsuk-form-group"
            >
              <label
                class="nhsuk-label nhsuk-date-input__label"
                for="dateinput_:r1a:-month"
                id="dateinput_:r1a:-month--label"
              >
                Month
              </label>
              <input
                aria-labelledby="dateinput_:r1a:-month--label"
                as="input"
                class="nhsuk-input nhsuk-input--width-2 nhsuk-date-input__input nhsuk-input--error"
                id="dateinput_:r1a:-month"
                inputmode="numeric"
                name="dateinput_:r1a:-month"
                pattern="[0-9]*"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="nhsuk-date-input__item"
          >
            <div
              class="nhsuk-form-group"
            >
              <label
                class="nhsuk-label nhsuk-date-input__label"
                for="dateinput_:r1a:-year"
                id="dateinput_:r1a:-year--label"
              >
                Year
              </label>
              <input
                aria-labelledby="dateinput_:r1a:-year--label"
                as="input"
                class="nhsuk-input nhsuk-input--width-4 nhsuk-date-input__input nhsuk-input--error"
                id="dateinput_:r1a:-year"
                inputmode="numeric"
                name="dateinput_:r1a:-year"
                pattern="[0-9]*"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`should render the date input component with an error object 1`] = `
<div>
  <div
    class="nhsuk-form-group nhsuk-form-group--error"
  >
    <div
      class="nhsuk-form-group"
    >
      <fieldset
        aria-describedby="dateinput_:r1o:_:r1p:--hint dateinput_:r1o:--error-message"
        aria-labelledby="dateinput_:r1o:--label"
        class="nhsuk-fieldset"
        role="group"
      >
        <legend
          class="nhsuk-legend"
          for="dateinput_:r1o:"
          id="dateinput_:r1o:--label"
        >
          Date of birth
        </legend>
        <div
          class="nhsuk-hint"
          id="dateinput_:r1o:_:r1p:--hint"
        >
          For example, 31 3 1980
        </div>
        <span
          class="nhsuk-error-message"
          id="dateinput_:r1o:--error-message"
        >
          <span
            class="nhsuk-u-visually-hidden"
          >
            Error: 
          </span>
          Please enter a valid day. Please enter a valid month.
        </span>
        <div
          class="nhsuk-date-input"
          id="dateinput_:r1o:"
        >
          <div
            class="nhsuk-date-input__item"
          >
            <div
              class="nhsuk-form-group"
            >
              <label
                class="nhsuk-label nhsuk-date-input__label"
                for="dateinput_:r1o:-day"
                id="dateinput_:r1o:-day--label"
              >
                Day
              </label>
              <input
                aria-labelledby="dateinput_:r1o:-day--label"
                as="input"
                class="nhsuk-input nhsuk-input--width-2 nhsuk-date-input__input nhsuk-input--error"
                id="dateinput_:r1o:-day"
                inputmode="numeric"
                name="dateinput_:r1o:-day"
                pattern="[0-9]*"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="nhsuk-date-input__item"
          >
            <div
              class="nhsuk-form-group"
            >
              <label
                class="nhsuk-label nhsuk-date-input__label"
                for="dateinput_:r1o:-month"
                id="dateinput_:r1o:-month--label"
              >
                Month
              </label>
              <input
                aria-labelledby="dateinput_:r1o:-month--label"
                as="input"
                class="nhsuk-input nhsuk-input--width-2 nhsuk-date-input__input nhsuk-input--error"
                id="dateinput_:r1o:-month"
                inputmode="numeric"
                name="dateinput_:r1o:-month"
                pattern="[0-9]*"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="nhsuk-date-input__item"
          >
            <div
              class="nhsuk-form-group"
            >
              <label
                class="nhsuk-label nhsuk-date-input__label"
                for="dateinput_:r1o:-year"
                id="dateinput_:r1o:-year--label"
              >
                Year
              </label>
              <input
                aria-labelledby="dateinput_:r1o:-year--label"
                as="input"
                class="nhsuk-input nhsuk-input--width-4 nhsuk-date-input__input"
                id="dateinput_:r1o:-year"
                inputmode="numeric"
                name="dateinput_:r1o:-year"
                pattern="[0-9]*"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </fieldset>
    </div>
  </div>
</div>
`;
